### DISCUSSION TITLE
Running the program

### DISCUSSION THREAD

**Discussion by Gooner12** [anomaly/gallagher#57] [Q&A] [answered]
**Issue:**

Failed to run the program. I tried running the program with, 
```sh
poetry run gala alarm list
```
The error I got:
```sh
AttributeError: 'NoneType' object has no attribute 'href'
```
Logging the `get_config()` returned this object:
```sh
EndpointConfig(endpoint=None, dto_list=<class 'gallagher.dto.response.alarm.AlarmSummaryResponse'>, dto_retrieve=<class  
'gallagher.dto.detail.alarm.AlarmDetail'>, top=10, sort='id', fields=(), search=())
```
I identified that the issue in the code is from `cls.__config__.endpoint.href`. Since, endpoint is None, the error is shown.

I have not set the api key before running the program. I guess this might be because the api key is not set. 

**Question:**

@devraj do you think this might be related to the missing API? Also, how to get the api key `GACC_API_KEY`?

**Comment by devraj** [comment, marked as answer]
> **Issue:**
> 
> Failed to run the program. I tried running the program with,
> 
> ```shell
> poetry run gala alarm list
> ```
> 
> The error I got:
> 
> ```shell
> AttributeError: 'NoneType' object has no attribute 'href'
> ```
> 
> Logging the `get_config()` returned this object:
> 
> ```shell
> EndpointConfig(endpoint=None, dto_list=<class 'gallagher.dto.response.alarm.AlarmSummaryResponse'>, dto_retrieve=<class  
> 'gallagher.dto.detail.alarm.AlarmDetail'>, top=10, sort='id', fields=(), search=())
> ```
> 
> I identified that the issue in the code is from `cls.__config__.endpoint.href`. Since, endpoint is None, the error is shown.
> 

I am pretty certain it is due to the missing API Key. In saying that the completed utility should raise proper error messages for the end user. If you follow the `alpha-5` branch then this is something I am working to finish in the coming weeks.

The endpoints are `None` because they are dynamically discovered from the server you point the client to, so without an API key it would not be able to proceed.

> I have not set the api key before running the program. I guess this might be because the api key is not set.
> 
> **Question:**
> 
> @devraj do you think this might be related to the missing API? Also, how to get the api key `GACC_API_KEY`?

You require a Gallagher Command Centre to actually use the utility as its intention is an extension to the REST API that they expose. For this project we run the tests and utilities against a test command centre that we manage.

In the near future we intend to:
- Provide a hosted interface for users to play with via something like `textual-web`
- Ensure that our default command centre can periodically reset its state (since it's connected to real hardware)

A solution would be to create a `mock` server, but that requires a significant amount of effort.

**Reply by Gooner12** [reply]
The issue indeed is caused by the `API_KEY`. Upon logging the response while discovering the endpoint, I got `<Response [401 Unauthorized]>`. I was thinking that we should create an issue around that, but since you are working on that already that is good. 

Will look into creating a `mock` server to see if it is feasible.

Thank you @devraj !!

**Reply by devraj** [reply]
Glad to hear you were able to validate it. 

I would refrain from building the `mock` server just yet as there are some nuances around how their API behaves.

`alpha-5` will handle the error messages a lot better on the cli.
